ddprofiler/app/src/main/java/ddprofiler/preanalysis/PreAnalyzer.java

```diff
@@ -130,26 +150,100 @@ public Map<Attribute, Values> readRows(int num) {
                 continue;
-            AttributeType aType;
+
+            AttributeType attributeType;
             if (pc.getBoolean(ProfilerConfig.EXPERIMENTAL)) {
                 // In experimental mode - force all data to be strings
-                aType = AttributeType.STRING;
+                attributeType = AttributeType.STRING;
             } else {
-                aType = typeOfValue(e.getValue());
+                attributeType = typeOfValue(entry.getValue());
             }
-            if (aType == null) {
+            if (attributeType == null) {
                 continue; // Means that data was dirty/anomaly, so skip value
             }
-            a.setColumnType(aType);
-            // a.setColumnType(AttributeType.STRING);
+            attribute.setColumnType(attributeType);
         }
     }
 
+    private void estimateSemanticTypes(Map<Attribute, List<String>> data) {
+        // estimate data type for each attribute
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+            if (attribute.getSemanticType() != null && !attribute.getSemanticType().isEmpty()) {
+                continue;
+            }
+            Map<String, String> semanticType = semanticTypeOfValue(entry.getValue());
+            attribute.setSemanticType(semanticType);
+        }
+    }
+
+    private Map<String, String> semanticTypeOfValue(List<String> values) {
+        Map<String, String> semanticTypes = new HashMap<>();
+        Map<String, Integer> temporalMatchCounts = new HashMap<>();
+        Map<String, Integer> spatialMatchCounts = new HashMap<>();
+
+        for (String value : values) {
+            String granularity = checkTemporalGranularity(value);
+            if (granularity != null) {
+                temporalMatchCounts.put(granularity, temporalMatchCounts.getOrDefault(granularity, 0) + 1);
+                continue;
+            }
+
+            granularity = checkSpatialGranularity(value);
+            if (granularity != null) {
+                spatialMatchCounts.put(granularity, spatialMatchCounts.getOrDefault(granularity, 0) + 1);
+            }
+        }
+
+        if (temporalMatchCounts.isEmpty() && spatialMatchCounts.isEmpty()) {
+            return semanticTypes;
+        }
+
+        // The dummy entry is created to avoid having to check which map is empty
+        Entry<String, Integer> maxTemporalGranularityEntry = temporalMatchCounts.entrySet().stream()
+                .max(Entry.comparingByValue()).orElse(new AbstractMap.SimpleEntry<>("dummy", 0));
+        Entry<String, Integer> maxSpatialGranularityEntry = spatialMatchCounts.entrySet().stream()
+                .max(Entry.comparingByValue()).orElse(new AbstractMap.SimpleEntry<>("dummy", 0));
+
+        if (maxTemporalGranularityEntry.getValue() > maxSpatialGranularityEntry.getValue()) {
+            semanticTypes.put("type", "temporal");
```

Comment:
It would be better to maintain these string values such as "temporal", "spatial", and other granularity values (including "unknown") in an enum instead of embedding them in the code. Like the design of `AttributeType`. It would make the code much easier to maintain for future developers.

Reply:
For the spatio-temporal type, I agree that creating an enum (SPATIAL and TEMPORAL) may make sense, but if we made granularity an enum as well, it would be harder if users would like to add new granularity patterns for either spatial or temporal. If we were to change spatio-temporal type to enum, I think it also wouldn't make sense to have semanticType Map to store the type and granularity. It would probably be better to have a SpatioTemporalType type attribute and a separate String(?) granularity attribute. However, attribute granularity does not stand alone and is always related to the type attribute and if in the future, we want to change the types to other semantic types, we can only implement the changes by adding new attributes (which may also be related to the type attribute. This is why I thought having semanticType as a Map (which would then be serialized into JSON) would work here.

Reply:
that makes sense. What I meant is to create an enum for SPATIAL and TEMPORAL. We still store semanticType as a dictionary.  For example, Instead of {"type": "temporal", "granularity": "day"}, we do {"type": SemanticType.TEMPORAL, "granularity": "day"}. We don't need to store granularity as an Enum because granularity is defined in the regex reference file.

Reply:
Due to the way Java's Map (dictionary) works, we can't have the key "type" have an enum (non-string) value, while key "granularity" have a string value. We could convert the type to string when storing in the dictionary, but I think it would defeat of the purpose of having it as enum in the first place.

I store information about an attribute's spatial/temporal-ness in `columnSemanticType` while having `columnSemanticTypeDetails` dictionary contain information such as granularity here [b781b1a](https://github.com/TheDataStation/ver/pull/71/commits/b781b1ab919d66ae8694f435890fd7fb3632d40a).

Reply:
That's great!